mutation_reader.cc

```diff
@@ -1216,11 +1219,25 @@ future<const mutation_fragment*> shard_reader::remote_reader::ensure_buffer_cont
         if (reader.is_buffer_empty()) {
             return reader.is_end_of_stream();
         }
-        const auto& next_pos = reader.peek_buffer().position();
-        if (next_pos.region() != partition_region::clustered) {
+        // If we hit a non clustering position, we are good as progress
+        // is guaranteed (non clustering positions are *always* strictly
+        // monotonically increasing compared to neighbouring positions.
+        if (buffer.back().position().region() != partition_region::clustered) {
             return true;
         }
-        return !next_pos.key().equal(*_schema, buffer.back().position().key());
+        const auto& next_pos = reader.peek_buffer().position();
+        const auto res = _tri_cmp(buffer.back().position(), _position_in_partition);
+        // To ensure progress we have to ensure three tings:
+        // * We have seen a fragment with position that is *greater* the last
+        //   seen one from the last `fill_buffer()` call.
+        // * The first unconsumed fragment in the reader has a position that is
+        //   *greater* than the last fragment in the buffer. This means that we
+        //   are guaranteed to have seen all fragment for the position of the
+        //   last fragment in the buffer and therefore it will be safe to
+        //   continue *after* that position in the next `fill_buffer()` call
+        //   (if the reader is evicted in the meanwhile).
```

Comment:
If the second point is satisfied, do we still need to check the first? Isn't it also guaranteed?

Reply:
We have to ensure:
```
_next_position_in_partition < buffer.back_position() < reader.peek_buffer().position()
```
Ensuring either condition won't imply the other.